presto-main/src/main/java/com/facebook/presto/memory/ClusterMemoryManager.java

```diff
@@ -264,14 +271,16 @@ public synchronized void process(Iterable<QueryExecution> runningQueries, Suppli
                     query.fail(exceededGlobalUserLimit(succinctBytes(userMemoryLimit)));
                     queryKilled = true;
                 }
-
-                long totalMemoryLimit = min(maxQueryTotalMemory.toBytes(), getQueryMaxTotalMemory(query.getSession()).toBytes());
-                if (totalMemoryReservation > totalMemoryLimit) {
-                    query.fail(exceededGlobalTotalLimit(succinctBytes(totalMemoryLimit)));
+                Optional<QueryLimit<DataSize>> queryTotalMemoryLimit = getMinimum(
+                        createDataSizeLimit(maxQueryTotalMemory, SYSTEM),
+                        query.getResourceGroupQueryLimits().flatMap(ResourceGroupQueryLimits::getTotalMemoryLimit).map(
```

Comment:
I'll stop commenting with these suggestions, but please make sure these new streams have each part of the stream in a newline, as suggested below, for readability.

```suggestion
                        query.getResourceGroupQueryLimits()
                                .flatMap(ResourceGroupQueryLimits::getTotalMemoryLimit)
                                .map(
```